src/permission/permission.h

```diff
@@ -0,0 +1,86 @@
+#ifndef SRC_PERMISSION_PERMISSION_H_
+#define SRC_PERMISSION_PERMISSION_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include "node_options.h"
+#include "permission/child_process_permission.h"
+#include "permission/fs_permission.h"
+#include "permission/permission_node.h"
+#include "permission/worker_permission.h"
+#include "v8.h"
+
+#include <iostream>
+#include <map>
+
+namespace node {
+
+class Environment;
+
+namespace permission {
+
+#define THROW_IF_INSUFFICIENT_PERMISSIONS(env, perm_, resource_, ...)          \
+  do {                                                                         \
+    if (UNLIKELY(!(env)->permission()->is_granted(perm_, resource_))) {        \
+      node::permission::Permission::ThrowAccessDenied((env), perm_);           \
+      return __VA_ARGS__;                                                      \
+    }                                                                          \
+  } while (0)
+
+class Permission {
+ public:
+  Permission() : enabled_(false) {
+    std::shared_ptr<PermissionNode> fs = std::make_shared<FSPermission>();
+    std::shared_ptr<PermissionNode> child_p =
+        std::make_shared<ChildProcessPermission>();
+    std::shared_ptr<PermissionNode> worker_t =
+        std::make_shared<WorkerPermission>();
+#define V(Name, _, __)                                                         \
+  nodes_.insert(std::make_pair(PermissionScope::k##Name, fs));
+    FILESYSTEM_PERMISSIONS(V)
+#undef V
+#define V(Name, _, __)                                                         \
+  nodes_.insert(std::make_pair(PermissionScope::k##Name, child_p));
+    CHILD_PROCESS_PERMISSIONS(V)
+#undef V
+#define V(Name, _, __)                                                         \
+  nodes_.insert(std::make_pair(PermissionScope::k##Name, worker_t));
+    WORKER_THREADS_PERMISSIONS(V)
+#undef V
+  }
+
+  inline bool is_granted(const PermissionScope permission, const char* res) {
+    if (!enabled_) return true;
+    auto perm_node = nodes_.find(permission);
+    if (perm_node != nodes_.end()) {
+      return perm_node->second->is_granted(permission, res);
+    }
+    return false;
+  }
+
+  inline bool is_granted(const PermissionScope permission,
+                         const std::string& res = "") {
+    return is_granted(permission, res.c_str());
```

Comment:
is_granted is called directly by `node_file.cc`. `*BufferValue` returns a `char*`, so accepting char* was needed to avoid another overhead when calling it from fs functions.

Reply:
Gonna reopen this because:

> `*BufferValue` returns a `char*`, so accepting char* was needed to avoid another overhead when calling it from fs functions.

It doesn’t (just) return a `char*`, it returns a `char*` *and* a length. Only taking the `char*` risks missing `\0` bytes that are part of the string/buffer, which is something you definitely don’t want to do in security-sensitive code.

(It might be a good idea to accept a `std::string_view` overload though, and add `.ToStringView()` helpers to `BufferValue` and `Utf8Value` similar to the `.ToString()` helpers that already exist.)

Reply:
Good point. Fixed in e91a563879.